syntax = "proto3";

package mcc.fred.middleware;
option go_package = ".;middleware";

// This extends the Fred client for more convenient functions.
service Middleware {
  rpc CreateKeygroup (CreateKeygroupRequest) returns (Empty);
  rpc DeleteKeygroup (DeleteKeygroupRequest) returns (Empty);
  rpc Read (ReadRequest) returns (ReadResponse);
  rpc Scan (ScanRequest) returns (ScanResponse);
  rpc Update (UpdateRequest) returns (Empty);
  rpc Delete (DeleteRequest) returns (Empty);
  rpc Append (AppendRequest) returns (AppendResponse);
  rpc Notify (NotifyRequest) returns (NotifyResponse);
  rpc AddReplica (AddReplicaRequest) returns (Empty);
  rpc GetKeygroupInfo (GetKeygroupInfoRequest) returns (GetKeygroupInfoResponse);
  rpc RemoveReplica (RemoveReplicaRequest) returns (Empty);
  rpc GetReplica (GetReplicaRequest) returns (GetReplicaResponse);
  rpc GetAllReplica (GetAllReplicaRequest) returns (GetAllReplicaResponse);
  rpc GetKeygroupTriggers (GetKeygroupTriggerRequest) returns (GetKeygroupTriggerResponse);
  rpc AddTrigger (AddTriggerRequest) returns (Empty);
  rpc RemoveTrigger (RemoveTriggerRequest) returns (Empty);
  rpc AddUser (UserRequest) returns (Empty);
  rpc RemoveUser (UserRequest) returns (Empty);
}

enum UserRole {
  ReadKeygroup = 0 ;
  WriteKeygroup = 1;
  ConfigureReplica = 2;
  ConfigureTrigger = 3 ;
  ConfigureKeygroups = 4;
}

message Empty {}

message CreateKeygroupRequest {
  string keygroup = 1;
  bool mutable = 2;
  int64 expiry = 3;
  string firstNodeId = 4;
}

message DeleteKeygroupRequest {
  string keygroup = 1;
}

message Item  {
  string id = 1;
  string val = 2;
  map<string, uint64> version = 3;
}

message ReadRequest {
  string keygroup = 1;
  string id = 2;
  int64 minExpiry = 3;
}

message ReadResponse {
  repeated Item items = 1;
}

message ScanRequest {
  string keygroup = 1;
  string id = 2;
  uint64 count = 3;
}

message ScanResponse {
  repeated Data data = 1;
}

message Data {
  string id = 1;
  string data = 2;
}

message UpdateRequest {
  string keygroup = 1;
  string id = 2;
  string data = 3;
}

message AppendRequest {
  string keygroup = 1;
  string data = 2;
}

message AppendResponse {
  string id = 1;
}

message NotifyRequest{
  string keygroup = 1;
  string id = 2;
  map<string, uint64> version = 3;
}

message NotifyResponse{}

message DeleteRequest {
  string keygroup = 1;
  string id = 2;
}

message AddReplicaRequest {
  string keygroup = 1;
  string nodeId = 2;
  int64 expiry = 3;
}

message GetKeygroupInfoRequest {
  string keygroup = 1;
}

message GetKeygroupInfoResponse {
  bool mutable = 1;
  repeated KeygroupReplica replica = 2;
}

message KeygroupReplica {
  string nodeId = 1;
  int64 expiry = 2;
  string host = 3;
}

message RemoveReplicaRequest {
  string keygroup = 1;
  string nodeId = 2;
}

message GetReplicaRequest {
  string nodeId = 1;
}

message GetReplicaResponse {
  string nodeId = 1;
  string host = 2;
}

message GetAllReplicaRequest {

}

message GetAllReplicaResponse {
  repeated GetReplicaResponse replicas = 1;
}

message GetKeygroupTriggerRequest {
  string keygroup = 1;
}

message GetKeygroupTriggerResponse {
  repeated Trigger triggers = 1;
}

message Trigger {
  string id = 1;
  string host = 2;
}

message AddTriggerRequest {
  string keygroup = 1;
  string triggerId = 2;
  string triggerHost = 3;
}

message RemoveTriggerRequest {
  string keygroup = 1;
  string triggerId = 2;
}

message UserRequest {
  string user = 1;
  string keygroup = 2;
  UserRole role = 3;
}